samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -173,71 +226,151 @@ public void deleteStartpoint(SystemStreamPartition ssp, TaskName taskName) {
+      if (CollectionUtils.isEmpty(ssps)) {
+        LOG.warn("No SSPs are mapped to taskName: {}", taskName.getTaskName());
+        continue;
+      }
+      for (SystemStreamPartition ssp : ssps) {
+        Startpoint startpoint = readStartpoint(ssp); // Read SSP-only key
+        if (startpoint != null) {
+          deleteKeys.put(ssp, null);
+        }
+
+        Startpoint startpointForTask = readStartpoint(ssp, taskName); // Read SSP+taskName key
+        if (startpointForTask != null) {
+          deleteKeys.put(ssp, taskName);
+        }
+
+        Startpoint startpointWithPrecedence = resolveStartpointPrecendence(startpoint, startpointForTask);
+        if (startpointWithPrecedence == null) {
+          continue;
+        }
+
+        fanOuts.putIfAbsent(taskName, new StartpointFanOut(now, new HashMap<>()));
+        fanOuts.get(taskName).getFanOuts().put(ssp, startpointWithPrecedence);
+      }
+    }
+
+    if (fanOuts.isEmpty()) {
+      LOG.debug("No fan outs created.");
+      return ImmutableSet.of();
+    }
+
+    LOG.info("Fanning out to {} tasks", fanOuts.size());
+
+    StartpointFanOut.StartpointFanOutSerde startpointFanOutSerde = new StartpointFanOut.StartpointFanOutSerde();
+    for (TaskName taskName : fanOuts.keySet()) {
+      String fanOutKey = toFanOutStoreKey(taskName);
+      StartpointFanOut newFanOut = fanOuts.get(taskName);
+      byte[] fanOutFromStoreInBytes = fanOutStore.get(fanOutKey);
+
+      if (ArrayUtils.isNotEmpty(fanOutFromStoreInBytes)) {
+        // Merge new fan out with existing fan out. This is not a typical scenario but it can happen when the job
+        // restarts before all startpoints are committed by the first checkpoint commit and new startpoints were written
+        // to the readWriteStore after the previous fan out.
+        StartpointFanOut existingFanOut = startpointFanOutSerde.fromBytes(fanOutFromStoreInBytes);
+        existingFanOut.getFanOuts().putAll(newFanOut.getFanOuts());
+        newFanOut = existingFanOut;
+      }
+
+      fanOutStore.put(fanOutKey, startpointFanOutSerde.toBytes(newFanOut));
 
+      // Delete from readWriteStore after fan out.
+      for (SystemStreamPartition ssp : newFanOut.getFanOuts().keySet()) {
+        if (deleteKeys.remove(ssp, taskName)) {
+          deleteStartpoint(ssp, taskName);
+        }
+        if (deleteKeys.remove(ssp, null)) {
+          deleteStartpoint(ssp);
         }
       }
     }
 
-    // Delete SSP-only keys
-    sspsToDelete.forEach(ssp -> {
-        deleteStartpoint(ssp);
-        LOG.info("All Startpoints for SSP: {} have been grouped to the appropriate tasks and the SSP was deleted.");
-      });
+    return ImmutableSet.copyOf(fanOuts.keySet());
+  }
+
+  /**
+   * Read the fanned out {@link Startpoint}s for the given {@link TaskName}
+   * @param taskName to read the fan out Startpoints for
+   * @return fanned out Startpoints
+   */
+  public Map<SystemStreamPartition, Startpoint> getFanOutForTask(TaskName taskName) {
```

Comment:
I thought about this as well. The read/write to non-fanout is sort of the public facing API of startpoints and the read/write to fanout store is entirely internal. I agree that separating these two categories into 2 classes makes the responsibilities more explicit.
The counter argument in having it all in one place is that it gives a better bird's eye view of the full startpoint flow.
If I were to create two classes, any suggestions for naming the non-fanout class? The fanout class is obvious of course.

Reply:
Makes sense that it is nice to see the whole flow in one class, but it does still seem like there are at least 2 distinct flows: 1) someone creates a startpoint from outside Samza, 2) Samza needs to interpret and use the startpoint
The only naming suggestion that comes to mind is `StartpointClient`, or something like that. I'm not very tied to it though.

Reply:
Also if we were to separate into two classes, the fan out class would depend on the readWrite class to move the startpoints from the readWriteStore to the fanOutStore. Since this refactor would be significant, I'll do it in a follow on PR.